src/charm.py

```diff
@@ -293,8 +292,8 @@ def _get_redis_relation_data(self) -> typing.Tuple[str, int]:
         for redis_unit in self._stored.redis_relation:  # type: ignore
             # mypy fails to see that this is indexable
             redis_unit_data = self._stored.redis_relation[redis_unit]  # type: ignore
-            redis_hostname = redis_unit_data.get("hostname", "")  # type: ignore
-            redis_port = redis_unit_data.get("port", DEFAULT_REDIS_PORT)  # type: ignore
+            redis_hostname = str(redis_unit_data.get("hostname", "") or "")
```

Comment:
Should this raise an error if the expected data is not present? Returning an empty string when it isn't could be misleading to the caller. Same for the port

Reply:
This function is always returning a tuple of a string and an integer for hostname and port. Later on in the code we're validating if this sufficient to consider the relation "ready" as follows: `if not all(self._get_redis_relation_data()):`. Is there a particular reason you think it's better to validate the data here/raise an error if it's not complete?

Reply:
yes because the data missing is essentially an error which can be caught with a `try ... except`. This way, it is explicit that something is wrong rather than implicit based on the value

Reply:
I changed the code to test if errors are raised outside of `get_redis_relation_data()` (following David's proposal).  
I decided to not do those checks in `_create_discourse_environment_settings()` because the data *should* be okay there (if it's not, then we end in error state but that is to be expected anyway).

Reply:
I also added a custom exception for the cases when a ValueError|TypeError is raised after discussion with David.